### PR TITLE
Custom album covers

### PR DESCRIPTION
**Pull request by evoludolab**
**Draft** pull request to implement setting custom cover images for albums, see LycheeOrg/Lychee#442

This pull request implements the front end for setting custom album covers by extends the context menu for photos and albums with _Set/Remove Album cover_ to set/remove custom album covers of the parent album. The corresponding changes in the back end are in a separate pull request LycheeOrg/Lychee#875

**IMPORTANT:**
* ⚠️ this draft uses `album.description` to store the information of custom covers. It requires a new column `cover` in the `album` table, which is beyond my pay grade.
* SQL query in `AlbumController.php` should be reviewed (this is completely unfamiliar territory...)
* something seems amiss with the gulpfile because iconic.svg does not get processed (the new icon needs to be manually added to resources/views/includes/svg.blade.php and resources/views/view.blade.php)
* only English localization done
* support for 3 distinct images for albums dropped (seems overkill given that 2 are barely visible); if this is acceptable some code cleaning and streamlining would be recommended

### CODE REVIEW

File: scripts/main/contextMenu.js [hunk 1]
```diff
@@ -209,13 +220,16 @@ contextMenu.albumTitle = function (albumID, e) {
...
 contextMenu.photo = function (photoID, e) {
-	// Notice for 'Move':
-	// fn must call basicContext.close() first,
-	// in order to keep the selection
+	let coverActive = photoID === album.json.cover_id;
 
 	let items = [
 		{ title: build.iconic("star") + lychee.locale["STAR"], fn: () => photo.setStar([photoID]) },
 		{ title: build.iconic("tag") + lychee.locale["TAGS"], fn: () => photo.editTags([photoID]) },
+		// for future work, use a list of all the ancestors.
```
Comment (kamil4) [inline scripts/main/contextMenu.js:228]: Is that actually needed? Selecting the cover one level at a time doesn't strike me as overly burdensome...

**Review by kamil4 — APPROVED** [review]
Minor comments overall. Still need to try it...

File: scripts/main/albums.js [hunk 2]
```diff
@@ -81,8 +81,18 @@ albums.load = function () {
 albums.parse = function (album) {
 	let i;
 	for (i = 0; i < 3; i++) {
-		if (!album.thumbs[i]) {
-			album.thumbs[i] = album.password === "1" ? "img/password.svg" : "img/no_images.svg";
+		if (lychee.api_V2) {
```
[the remark this answers is not in the document]
Comment (ildyria) [reply scripts/main/albums.js]: > Is the current front end still supposed to work with v3?

Honestly, I don't really care. But all the v3 things should be removed in a Single PR.

File: scripts/main/build.js [hunk 3]
```diff
@@ -127,6 +167,7 @@ build.photo = function (data, disabled = false) {
 	let html = "";
 	let thumbnail = "";
 	var thumb2x = "";
+	let isCover = data.id === album.json.cover_id;
```
[the remark this answers is not in the document]
Comment (ildyria) [reply scripts/main/build.js:140]: > I forget; are there cases when album.json might be null? (e.g., test what happens if you reload the browser window while in photo view)

then this menu is not accessible if I am not wrong.

File: scripts/main/contextMenu.js (hunk 1 above)
Comment (ildyria) [reply scripts/main/contextMenu.js:228]: Assume that you have an album without pictures but just sub albums...

File: scripts/main/build.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kamil4) [reply scripts/main/build.js:140]: Right...

File: scripts/main/contextMenu.js (hunk 1 above)
Comment (kamil4) [reply scripts/main/contextMenu.js:228]: Actually, selecting cover of a subalbum is currently broken. I'm fixing it...

**Comment by kamil4** [comment]
OK, I added the missing pieces for selecting a subalbum as cover; can you guys test?

Smart albums now display "public" badge for if the Starred or Recent albums are public. I'm undecided if I like that or not... It's probably an improvement, although I may tweak the ordering...

**Comment by ildyria** [comment]
I'm unsure about having the cover badge showing up for albums. For pictures it is pretty obvious but for album I feel this is overkill. :thinking:

**Comment by kamil4** [comment]
> I'm unsure about having the cover badge showing up for albums. For pictures it is pretty obvious but for album I feel this is overkill. thinking

I don't get it. To me they seem like exactly the same case? A user can select for a cover of an album either a photo from the album as or a cover from a subalbum. In either case we give the same visual indication. Why should one be obvious and the other an overkill?

**Comment by coq-verif-tweetnacl** [comment]
Let me check it on my server :)

**Comment by ildyria** [comment]
I will check. :)

**Comment by evoludolab** [comment]
In my original attempt only the context menu entry and the colour of its accompanying icon changed for a currently selected photo/sub album. This was subtle to the point of obscurity... I am with @kamil4 and already got totally used to the badges. My initial concerns with them completely evaporated after @ildyria pointed out that badges are only visible for the owner of the album.

**Pull request merged**